File: python/trinitylake/trinity_lake.py
import time
import uuid
from dataclasses import dataclass

from const import HIGHEST_CHARACTER_VALUE
from trinitylake.storage import Storage
from trinitylake.protobuf import LakehouseDef, NamespaceDef, TableDef
from trinitylake.storage import file_paths
from trinitylake.tree import TrinityTree, TrinityNode, tree_keys
from trinitylake.const import (
    LAKEHOUSE_DEF,
    VERSION,
    LATEST_HINT_FILE,
    NAMESPACE_SCHEMA_ID_PART,
    TABLE_SCHEMA_ID_PART,
)
from trinitylake.storage import tree_operations


@dataclass
class RunningTransaction:
    transaction_id: str
    beginning_tree: TrinityTree
    running_tree: TrinityTree
    began_at_millis: int


def create_lakehouse(storage: Storage, lakehouse_def: LakehouseDef):
    lakehouse_def_path = file_paths.new_lakehouse_def_path()
    storage.serialize_protobuf(lakehouse_def, lakehouse_def_path)

    root_node = TrinityNode()
    root_file_path = file_paths.get_root_node_path(0)
    root_node.system_rows.append((LAKEHOUSE_DEF, lakehouse_def_path))
    root_node.system_rows.append((VERSION, "0"))
    storage.write_node(root_node, root_file_path)

    storage.write_file(LATEST_HINT_FILE, str(0).encode("utf-8"))


def begin_transaction(storage: Storage):
    tree = tree_operations.find_latest_root(storage)

    return RunningTransaction(
        transaction_id=str(uuid.uuid4()),
        beginning_tree=tree,
        running_tree=tree,
        began_at_millis=int(time.time() * 1000),
    )


def commit_transaction(storage: Storage, transaction: RunningTransaction):
    version = int(_find_version(transaction.running_tree))
    root_node = transaction.running_tree.root
    new_version = version + 1
    # for change in changes:
    #     if not change[1]:
    #         trinity_tree.delete(change[0])
    #     else:
    #         trinity_tree.insert(change[0], change[1])
    # TODO: update system rows
    tree_name = file_paths.get_root_node_path(new_version)
    storage.serialize_tree(root_node, tree_name)
    storage.write_file(LATEST_HINT_FILE, str(new_version).encode("utf-8"))


def show_namespaces(transaction: RunningTransaction):
    running_root = transaction.running_tree
    return [
        tree_keys.get_namespace_name_from_key(k)
        for k, v in running_root.range_query(
            NAMESPACE_SCHEMA_ID_PART, NAMESPACE_SCHEMA_ID_PART + HIGHEST_CHARACTER_VALUE
        )
    ]


def create_namespace(
    storage: Storage,
    transaction: RunningTransaction,
    namespace_name: str,
    namespace_def: NamespaceDef,
):
    lakehouse_def = _find_lakehouse_def(storage, transaction.running_tree)
    namespace_key = tree_keys.get_namespace_key(namespace_name, lakehouse_def)
    try:
        transaction.running_tree.get(namespace_key)
    except KeyError:
        pass
    file_path = storage.serialize_protobuf(namespace_def)
    transaction.running_tree.insert(namespace_key, file_path)
    return transaction


def describe_namespace(
    storage: Storage, transaction: RunningTransaction, namespace_name: str
):
    lakehouse_def = _find_lakehouse_def(storage, transaction.running_tree)
    namespace_key = tree_keys.get_namespace_key(namespace_name, lakehouse_def)

    try:
        namespace_def = transaction.running_tree.get(namespace_key)
    except KeyError:
        raise ValueError(f"Namespace {namespace_name} does not exist")
    return storage.deserialize_protobuf(namespace_def, NamespaceDef)


def drop_namespace(
    storage: Storage, transaction: RunningTransaction, namespace_name: str
):
    lakehouse_def = _find_lakehouse_def(storage, transaction.running_tree)
    namespace_key = tree_keys.get_namespace_key(namespace_name, lakehouse_def)
    try:
        transaction.running_tree.get(namespace_key)
    except KeyError:
        raise ValueError(f"Namespace {namespace_name} does not exist")
    transaction.running_tree.delete(namespace_key)


def show_tables(transaction: RunningTransaction, namespace_name: str):
    running_root = transaction.running_tree
    start = f"{TABLE_SCHEMA_ID_PART}{namespace_name}"
    end = f"{start} {HIGHEST_CHARACTER_VALUE}"
    return [
        tree_keys.get_table_name_from_key(namespace_name, k)
        for k, v in running_root.range_query(start, end)
    ]


def describe_table(
    storage: Storage,
    transaction: RunningTransaction,
    namespace_name: str,
    table_name: str,
):
    lakehouse_def = _find_lakehouse_def(storage, transaction.running_tree)
    table_key = tree_keys.get_table_key(namespace_name, table_name, lakehouse_def)
    try:
        table_def = transaction.running_tree.get(table_key)
    except KeyError:
        raise ValueError(f"Table {table_name} does not exist")
    return storage.deserialize_protobuf(table_def, TableDef)


def create_table(
    storage: Storage,
    transaction: RunningTransaction,
    namespace_name: str,
    table_name: str,
    table_def: TableDef,
):
    lakehouse_def = _find_lakehouse_def(storage, transaction.running_tree)
    table_key = tree_keys.get_table_key(namespace_name, table_name, lakehouse_def)
    try:
        transaction.running_tree.get(table_key)
        raise ValueError(f"Table {table_name} already exists")
    except KeyError:
        pass
    file_path = storage.serialize_protobuf(table_def)
    transaction.running_tree.insert(table_key, file_path)
    return transaction


def _find_lakehouse_def(storage: Storage, tree: TrinityTree):
    for sys_key in tree.root.system_rows:
        if sys_key[0] == LAKEHOUSE_DEF:
            return storage.deserialize_protobuf(sys_key[1], LakehouseDef)
    raise ValueError("No lakehouse definition found")


def _find_version(tree: TrinityTree):
    for sys_key in tree.root.system_rows:
        if sys_key[0] == VERSION:
            return sys_key[1]
    raise ValueError("No version found")

File: core/src/main/java/io/trinitylake/TrinityLake.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.trinitylake;

import io.trinitylake.exception.CommitFailureException;
import io.trinitylake.exception.ObjectAlreadyExistsException;
import io.trinitylake.exception.ObjectNotFoundException;
import io.trinitylake.models.LakehouseDef;
import io.trinitylake.models.NamespaceDef;
import io.trinitylake.models.TableDef;
import io.trinitylake.models.ViewDef;
import io.trinitylake.relocated.com.google.common.collect.ImmutableMap;
import io.trinitylake.storage.LakehouseStorage;
import io.trinitylake.tree.BasicTreeRoot;
import io.trinitylake.tree.NodeKeyTableRow;
import io.trinitylake.tree.TreeOperations;
import io.trinitylake.tree.TreeRoot;
import io.trinitylake.util.ValidationUtil;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public class TrinityLake {

  private TrinityLake() {}

  public static void createLakehouse(LakehouseStorage storage, LakehouseDef lakehouseDef) {
    String lakehouseDefFilePath = FileLocations.newLakehouseDefFilePath();
    ObjectDefinitions.writeLakehouseDef(storage, lakehouseDefFilePath, lakehouseDef);

    BasicTreeRoot root = new BasicTreeRoot();
    root.setLakehouseDefFilePath(lakehouseDefFilePath);
    String rootNodeFilePath = FileLocations.rootNodeFilePath(0);
    TreeOperations.writeRootNodeFile(storage, rootNodeFilePath, root);
    TreeOperations.tryWriteRootNodeVersionHintFile(storage, 0);
  }

  public static RunningTransaction beginTransaction(LakehouseStorage storage) {
    return beginTransaction(storage, ImmutableMap.of());
  }

  public static RunningTransaction beginTransaction(
      LakehouseStorage storage, Map<String, String> options) {
    TreeRoot current = TreeOperations.findLatestRoot(storage);
    TransactionOptions transactionOptions = new TransactionOptions(options);
    return ImmutableRunningTransaction.builder()
        .beganAtMillis(System.currentTimeMillis())
        .transactionId(UUID.randomUUID().toString())
        .beginningRoot(current)
        .runningRoot(current)
        .isolationLevel(transactionOptions.isolationLevel())
        .build();
  }

  public static CommittedTransaction commitTransaction(
      LakehouseStorage storage, RunningTransaction transaction) throws CommitFailureException {
    ValidationUtil.checkArgument(
        !transaction.runningRoot().path().isPresent(), "There is no change to be committed");
    ValidationUtil.checkState(
        transaction.beginningRoot().path().isPresent(),
        "Cannot find persisted storage path for beginning root");

    String beginningRootNodeFilePath = transaction.beginningRoot().path().get();
    long beginningRootVersion = FileLocations.versionFromNodeFilePath(beginningRootNodeFilePath);
    long nextRootVersion = beginningRootVersion + 1;
    String nextVersionFilePath = FileLocations.rootNodeFilePath(nextRootVersion);
    transaction.runningRoot().setPreviousRootNodeFilePath(beginningRootNodeFilePath);

    TreeOperations.writeRootNodeFile(storage, nextVersionFilePath, transaction.runningRoot());
    TreeOperations.tryWriteRootNodeVersionHintFile(storage, nextRootVersion);
    transaction.runningRoot().setPath(nextVersionFilePath);
    return ImmutableCommittedTransaction.builder()
        .committedRoot(transaction.runningRoot())
        .transactionId(transaction.transactionId())
        .build();
  }

  public static List<String> showNamespaces(
      LakehouseStorage storage, RunningTransaction transaction) {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    return transaction.runningRoot().nodeKeyTable().stream()
        .map(NodeKeyTableRow::key)
        .filter(key -> ObjectKeys.isNamespaceKey(key, lakehouseDef))
        .map(key -> ObjectKeys.namespaceNameFromKey(key, lakehouseDef))
        .collect(Collectors.toList());
  }

  public static boolean namespaceExists(
      LakehouseStorage storage, RunningTransaction transaction, String namespaceName) {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    return TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent();
  }

  public static NamespaceDef describeNamespace(
      LakehouseStorage storage, RunningTransaction transaction, String namespaceName)
      throws ObjectNotFoundException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    Optional<String> namespaceDefFilePath =
        TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey);
    if (!namespaceDefFilePath.isPresent()) {
      throw new ObjectNotFoundException("Namespace %s does not exist", namespaceName);
    }
    return ObjectDefinitions.readNamespaceDef(storage, namespaceDefFilePath.get());
  }

  public static RunningTransaction createNamespace(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      NamespaceDef namespaceDef)
      throws ObjectAlreadyExistsException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    if (TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent()) {
      throw new ObjectAlreadyExistsException("Namespace %s already exists", namespaceName);
    }

    String namespaceDefFilePath = FileLocations.newNamespaceDefFilePath(namespaceName);
    ObjectDefinitions.writeNamespaceDef(storage, namespaceDefFilePath, namespaceName, namespaceDef);
    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.setValue(storage, newRoot, namespaceKey, namespaceDefFilePath);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static RunningTransaction alterNamespace(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      NamespaceDef namespaceDef)
      throws ObjectNotFoundException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent()) {
      throw new ObjectNotFoundException("Namespace %s does not exist", namespaceName);
    }

    String namespaceDefFilePath = FileLocations.newNamespaceDefFilePath(namespaceName);
    ObjectDefinitions.writeNamespaceDef(storage, namespaceDefFilePath, namespaceName, namespaceDef);
    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.setValue(storage, newRoot, namespaceKey, namespaceDefFilePath);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static RunningTransaction dropNamespace(
      LakehouseStorage storage, RunningTransaction transaction, String namespaceName)
      throws ObjectNotFoundException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent()) {
      throw new ObjectNotFoundException("Namespace %s does not exist", namespaceName);
    }

    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.removeKey(storage, newRoot, namespaceKey);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static List<String> showTables(
      LakehouseStorage storage, RunningTransaction transaction, String namespaceName)
      throws ObjectNotFoundException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    return transaction.runningRoot().nodeKeyTable().stream()
        .map(NodeKeyTableRow::key)
        .filter(key -> ObjectKeys.isTableKey(key, lakehouseDef))
        .map(key -> ObjectKeys.tableNameFromKey(key, lakehouseDef))
        .collect(Collectors.toList());
  }

  public static boolean tableExists(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String tableName) {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String tableKey = ObjectKeys.tableKey(namespaceName, tableName, lakehouseDef);
    return TreeOperations.searchValue(storage, transaction.runningRoot(), tableKey).isPresent();
  }

  public static TableDef describeTable(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String tableName)
      throws ObjectNotFoundException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String tableKey = ObjectKeys.tableKey(namespaceName, tableName, lakehouseDef);
    Optional<String> tableDefFilePath =
        TreeOperations.searchValue(storage, transaction.runningRoot(), tableKey);
    if (!tableDefFilePath.isPresent()) {
      throw new ObjectNotFoundException(
          "Namespace %s table %s does not exist", namespaceName, tableName);
    }
    return ObjectDefinitions.readTableDef(storage, tableDefFilePath.get());
  }

  public static RunningTransaction createTable(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String tableName,
      TableDef tableDef)
      throws ObjectAlreadyExistsException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent()) {
      throw new ObjectNotFoundException("Namespace %s does not exist", namespaceName);
    }
    String tableKey = ObjectKeys.tableKey(namespaceName, tableName, lakehouseDef);
    if (TreeOperations.searchValue(storage, transaction.runningRoot(), tableKey).isPresent()) {
      throw new ObjectAlreadyExistsException(
          "Namespace %s table %s already exists", namespaceName, tableName);
    }

    String tableDefFilePath = FileLocations.newTableDefFilePath(namespaceName, tableName);
    ObjectDefinitions.writeTableDef(storage, tableDefFilePath, namespaceName, tableName, tableDef);
    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.setValue(storage, newRoot, tableKey, tableDefFilePath);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static RunningTransaction alterTable(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String tableName,
      TableDef tableDef)
      throws ObjectNotFoundException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String namespaceKey = ObjectKeys.namespaceKey(namespaceName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), namespaceKey).isPresent()) {
      throw new ObjectNotFoundException("Namespace %s does not exist", namespaceName);
    }
    String tableKey = ObjectKeys.tableKey(namespaceName, tableName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), tableKey).isPresent()) {
      throw new ObjectNotFoundException(
          "Namespace %s table %s does not exists", namespaceName, tableName);
    }

    String tableDefFilePath = FileLocations.newTableDefFilePath(namespaceName, tableName);
    ObjectDefinitions.writeTableDef(storage, tableDefFilePath, namespaceName, tableName, tableDef);
    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.setValue(storage, newRoot, tableKey, tableDefFilePath);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static RunningTransaction dropTable(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String tableName)
      throws ObjectNotFoundException, CommitFailureException {
    LakehouseDef lakehouseDef = TreeOperations.findLakehouseDef(storage, transaction.runningRoot());
    String tableKey = ObjectKeys.tableKey(namespaceName, tableName, lakehouseDef);
    if (!TreeOperations.searchValue(storage, transaction.runningRoot(), tableKey).isPresent()) {
      throw new ObjectNotFoundException(
          "Namespace %s table %s does not exists", namespaceName, tableName);
    }

    TreeRoot newRoot = TreeOperations.cloneTreeRoot(transaction.runningRoot());
    TreeOperations.removeKey(storage, newRoot, tableKey);
    return ImmutableRunningTransaction.builder().from(transaction).runningRoot(newRoot).build();
  }

  public static List<String> showViews(
      LakehouseStorage storage, RunningTransaction transaction, String namespaceName) {
    return null;
  }

  public static boolean viewExists(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName) {
    return false;
  }

  public static ViewDef describeView(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName) {
    return null;
  }

  public static RunningTransaction createView(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName,
      TableDef viewDef) {
    return null;
  }

  public static RunningTransaction replaceView(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName,
      TableDef viewDef) {
    return null;
  }

  public static RunningTransaction alterView(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName,
      TableDef viewDef) {
    return null;
  }

  public static RunningTransaction dropView(
      LakehouseStorage storage,
      RunningTransaction transaction,
      String namespaceName,
      String viewName) {
    return null;
  }
}
